src/devices/Rtc/Devices/Ds3231/Ds3231.cs

```diff
@@ -99,6 +104,239 @@ protected double ReadTemperature()
+                NumberHelper.Bcd2Dec((byte)(rawData[2] & 0b_0111_1111)),
+                NumberHelper.Bcd2Dec((byte)(rawData[1] & 0b_0111_1111)),
+                NumberHelper.Bcd2Dec((byte)(rawData[0] & 0b_0111_1111)),
+                (Ds3231Alarm1MatchMode)matchMode);
+        }
+
+        /// <summary>
+        /// Sets alarm 1
+        /// </summary>
+        /// <param name="alarm">New alarm 1</param>
+        public void SetAlarm1(Ds3231Alarm1 alarm)
+        {
+            if (alarm == null)
+            {
+                throw new ArgumentNullException(nameof(alarm));
+            }
+
+            if (alarm.Second < 0 || alarm.Second > 59)
+            {
+                throw new ArgumentOutOfRangeException(nameof(alarm), "Second must be between 0 and 59.");
+            }
+
+            if (alarm.Minute < 0 || alarm.Minute > 59)
+            {
+                throw new ArgumentOutOfRangeException(nameof(alarm), "Minute must be between 0 and 59.");
+            }
+
+            if (alarm.Hour < 0 || alarm.Hour > 23)
+            {
+                throw new ArgumentOutOfRangeException(nameof(alarm), "Hour must be between 0 and 23.");
+            }
+
+            if (alarm.MatchMode == Ds3231Alarm1MatchMode.DayOfWeekHoursMinutesSeconds)
+            {
+                if (alarm.DayOfMonthOrWeek < 1 || alarm.DayOfMonthOrWeek > 7)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(alarm), "Day of week must be between 1 and 7.");
+                }
+            }
+            else if (alarm.MatchMode == Ds3231Alarm1MatchMode.DayOfMonthHoursMinutesSeconds)
+            {
+                if (alarm.DayOfMonthOrWeek < 1 || alarm.DayOfMonthOrWeek > 31)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(alarm), "Day of month must be between 1 and 31.");
+                }
+            }
+
+            Span<byte> setData = stackalloc byte[5];
+            setData[0] = (byte)Ds3231Register.RTC_ALM1_SEC_REG_ADDR;
+
+            setData[1] = NumberHelper.Dec2Bcd(alarm.Second);
+            setData[2] = NumberHelper.Dec2Bcd(alarm.Minute);
+            setData[3] = NumberHelper.Dec2Bcd(alarm.Hour);
+            setData[4] = NumberHelper.Dec2Bcd(alarm.DayOfMonthOrWeek);
+
+            setData[1] |= (byte)((byte)(((byte)alarm.MatchMode) & 1) << 7); // Set A1M1 bit
+            setData[2] |= (byte)((byte)(((byte)alarm.MatchMode) & (1 << 1)) << 6); // Set A1M2 bit
+            setData[3] |= (byte)((byte)(((byte)alarm.MatchMode) & (1 << 2)) << 5); // Set A1M3 bit
+            setData[4] |= (byte)((byte)(((byte)alarm.MatchMode) & (1 << 3)) << 4); // Set A1M4 bit
+            setData[4] |= (byte)((byte)(((byte)alarm.MatchMode) & (1 << 4)) << 2); // Set DY/DT bit
+
+            _i2cDevice.Write(setData);
+        }
+
+        /// <summary>
+        /// Reads the currently set alarm 2
+        /// </summary>
+        /// <returns>Alarm 1</returns>
+        public Ds3231Alarm2 ReadAlarm2()
```

Comment:
Consider: instead of Set/Read alarm methods have a nullable property and immutable class with a single instance of this class i.e.:

```csharp
Ds3231 x = ...;
x.Alarm = Ds3231.CreateAlarm(...); // you can have two similar methods for CreateAlarm to cover both cases (perhaps different name to better describe it, constructor is fine as well if args are different enough)
```